ipykernel/kernelbase.py

```diff
@@ -899,6 +915,41 @@ def _input_request(self, prompt, ident, parent, password=False):
             raise EOFError
         return value
 
+    def _wait_input_request_reply(self):
+        """Wait for an input request reply.
+
+        Raises
+        ------
+        KeyboardInterrupt if a keyboard interrupt is recieved.
+        """
+        # Await a response.
+        reply = None
+        while reply is None:
+            try:
+                reply = self._input_request_loop_step()
+            except Exception:
+                self.log.warning("Invalid Message:", exc_info=True)
+            except KeyboardInterrupt:
+                # re-raise KeyboardInterrupt, to truncate traceback
+                raise KeyboardInterrupt
+        return reply
+
+    def _input_request_loop_step(self):
+        """Do one step of the input request loop."""
+        # Allow GUI event loop to update
+        if sys.version_info >= (3, 4):
+            is_main_thread = (threading.current_thread() is
+                              threading.main_thread())
+        else:
+            is_main_thread = isinstance(threading.current_thread(),
+                                        threading._MainThread)
+        if is_main_thread and self.eventloop:
+            self.eventloop(self)
+            return self._stdin_msg
```

Comment:
and possibly here?

Reply:
`_input_request_loop_step` is called from `_wait_input_request_reply` which is protected by a lock in `_input_request`, so I don't think a lock here would do anything.